File: code/wctun1.f
      subroutine wctun1(iob,le,cexpo,cwhpid,cyear,cmonth,cday,stnnbr,
     *                  castno,nobs,instru,samphz,icd)
c
c -- routine to read and uncode the first 3 lines of a WOCE format high
c    resolution CTD data file. *.wct.
c -- David Newton, SIO apr94.
c
      integer iob,le,cyear,cmonth,cday,castno,nobs,icd
      real samphz
      character cexpo*14,cwhpid*5,stnnbr*6,instru*4
c
c   iob    ->  fortran unit for opened *.wct file. ready to read 1st line.
c   le     ->  fortran error unit.
c   cexpo  <-  14 char left justified expo code from the data file.
c   cwhpid <-  5 char left justified WOCE line (section ) id.
c   cyear  <-  2 digit year from data file.
c   cmonth <-  2 digit month from data file.
c   cday   <-  2 digit day from data file.
c   stnnbr <-  6 character station id from data file.
c   castno <-  castnumber (integer) from data file.
c   nobs   <-  number of high res levels in cast.
c   instru <-  4 character instrument id from data file.
c   samphz <-  sampling rate in hz. real.
c   icd    <-  error return code.
c              0==okay, 1==warning, 2==fatal err, negative=end_of_file
c
      character line*60,scrtch*20,fmt*20
      integer i,j,k,nsig
      logical warn
c
      intrinsic index
      external ljustn
c
      warn = .false.
c
c
c -- uncode the 1st header line.
c
      read(iob,2010,iostat=icd) line
 2010  format(a)
      if(icd .ne. 0) then
          write(le,*)'wctun1: read err on 1st line. iostat=',icd
          go to 300
      endif
c
      i = index(line,'EXPOCODE')
      if(i .eq. 0) then
          write(le,*)'wctun1: no EXPOCODE word found. line=',line
          go to 350
      endif
      j = index(line,'WHP')
      if(j .eq. 0) then
          write(le,*)'wctun1: no WHP-ID word found. line=',line
          go to 350
      endif
      if(i+8 .gt. j-1) then
          write(le,*)'wctun1: 1st header line screwed up. line=',line
          go to 350
      endif
      scrtch = line(i+8:j+1)
      call ljustn(scrtch,nsig)
      k = 14
      do 20 i=2,14
         if(scrtch(i:i) .eq. ' ') then
             k = i-1
             go to 21
         endif
 20   continue
 21   continue
      cexpo = scrtch(1:k)
c
      i = j
      j = index(line,'DATE')
      if(j .eq. 0) then
          write(le,*)'wctun1: no DATE word found. line=',line
          go to 350
      endif
      if(i+6 .gt. j-1) then
          write(le,*)'wctun1: 1st header line screwed up. line=',line
          go to 350
      endif
      scrtch = line(i+6:j-1)
      call ljustn(scrtch,nsig)
      cwhpid = scrtch
c
      scrtch = line(j+4:j+14)
      call ljustn(scrtch,nsig)
      if(scrtch(6:6) .eq. ' ') then
c         -- 5 char date.
          read(scrtch,'(i1,2i2)',iostat=icd) cmonth,cday,cyear
      else
c         -- 6 char date. (as it should be).
          read(scrtch,'(3i2)',iostat=icd) cmonth,cday,cyear
      endif
      if(icd .ne. 0 ) then
          write(le,*)'wctun1: read err on date. iostat,line=',icd,line
          warn = .true.
          cmonth = -9
          cday = -9
          cyear = -9
      endif
c
c -- uncode the second header line.
c
      read(iob,2010,iostat=icd) line
      if(icd .ne. 0) then
          write(le,*)'wctun1: read err on 2nd line. iostat=',icd
          go to 300
      endif
c
      i = index(line,'STNNBR')
      if(i .eq. 0) then
          write(le,*)'wctun1: no STNNBR word found. line=',line
          go to 350
      endif
      j = index(line,'CASTNO')
      if(j .eq. 0) then
          write(le,*)'wctun1: no CASTNO word found. line=',line
          go to 350
      endif
      if(i+6 .gt. j-1) then
          write(le,*)'wctun1: 2nd header line screwed up. line=',line
          go to 350
      endif
      scrtch = line(i+6:j-1)
      call ljustn(scrtch,nsig)
      stnnbr = scrtch
c
      i = j
      j = index(line,'NO.')
      if(j .eq. 0) then
          write(le,*)'wctun1: no NO. RECORDS= word found. line=',line
          go to 350
      endif
      k = (j-1) - (i+6) + 1
      if(k .lt. 1) then
          write(le,*)'wctun1: 2nd header line screwed up. line=',line
          go to 350
      endif
      write(fmt,2020,iostat=icd) k
 2020  format('(I',i2,')')
      if(icd .ne. 0) then
          write(le,*)'wctun1: err forming format for castno read.',
     *               'notify programmer. icd=',icd
          fmt = '(i4)'
      endif
      read(line(i+6:j-1),fmt,iostat=icd) castno
      if(icd .ne. 0) then
          write(le,*)'wctun1: read err on CASTNO. iostat,line=',
     *                icd,line
          go to 300
      endif
c
      i = index(line,'ORDS=')
      if(i .eq. 0) then
          write(le,*)'wctun1: no NO. RECORDS= word found. line=',line
          go to 350
      endif
      scrtch = line(i+5:i+5+10)
      call ljustn(scrtch,nsig)
      do 50 i=2,nsig+1
         if(scrtch(i:i) .eq. ' ') then
             k = i-1
             go to 51
         endif
 50   continue
      write(le,*)'wctun1: NO. RECORDS= unreadable. line=',line
      go to 350
 51   continue
      write(fmt,2020,iostat=icd) k
      if(icd .ne. 0) then
          write(le,*)'wctun1: err forming format for RECORDS= read.',
     *               'notify programmer. icd=',icd
          fmt = '(i5)'
      endif
      read(scrtch(1:k),fmt,iostat=icd) nobs
      if(icd .ne. 0) then
          write(le,*)'wctun1: read err on NO. RECORDS. iostat,line=',
     *                icd,line
          write(le,*)'       fmt=',fmt
          go to 300
      endif
c
c -- uncode 3rd header.
c
      read(iob,2010,iostat=icd) line
      if(icd .ne. 0) then
          write(le,*)'wctun1: read err on 2nd line. iostat=',icd
          go to 300
      endif
c
      i = index(line,'NO.')
      if(i .eq. 0) then
          write(le,*)'wctun1: no INSTRUMENT NO. word found. line=',
     *                line
          instru = '-'
          warn = .true.
          go to 210
      endif
      j =  index(line,'SAMP')
      if(j .eq. 0) then
          write(le,*)'wctun1: no SAMPLING RATE word found. line=',line
          instru = '-'
          warn = .true.
          go to 210
      endif
      if(i+3 .gt. j-1) then
          write(le,*)'wctun1: 3rd header line screwed up. line=',line
          instru = '-'
          warn = .true.
          go to 210
      endif
      scrtch = line(i+3:j-1)
      call ljustn(scrtch,nsig)
      instru = scrtch
      if(nsig .eq. 0) instru = '-'
c
 210  continue
      i = index(line,'RATE')
      if(i .eq. 0) then
          write(le,*)'wctun1: no SAMPLING RATE word found. line=',line
          samphz = -9.
          warn = .true.
          go to 220
      endif
      j = index(line,'HZ')
      if(j .eq. 0) then
          write(le,*)'wctun1: no HZ word found. line=',line
          samphz = -9.
          warn = .true.
          go to 220
      endif
      k = (j-1) - (i+4) + 1
      if(k .lt. 2) then
          write(le,*)'wctun1: can''t read sampling rate.'
          samphz = -9.
          warn = .true.
          go to 220
      endif
      write(fmt,2040,iostat=icd) k
 2040  format('(F',i2,'.2)')
      if(icd .ne. 0) then
          write(le,*)'wctun1: err forming format for hz. read.',
     *               'notify programmer. icd=',icd
          fmt = '(f7.2)'
      endif
      read(line(i+4:j-1),fmt,iostat=icd) samphz
      if(icd .ne. 0) then
          write(le,*)'wctun1: read err on HZ. iostat,line=',
     *                icd,line
          samphz = -9.
          warn = .true.
          go to 220
      endif
c
c -- get to here on everything normal and warnings only.
 220  continue
      if(warn) icd = 1
      return
c
c -- get to here on read error of some fatal type.
 300  if(icd .gt. 0) icd = 2
      return
c
c -- get to here on some fatal parsing error.
 350  icd = 2
      return
c
      end
  

File: src/wctun1.py
#!/usr/bin/env python

execfile('libwocecvt.py')
execfile('libwctcvt.py')

REQUIRED = ['EXPOCODE', 'WHP', 'DATE', 'STNNBR', 'CASTNO',
            'NO. RECORDS', 'INSTRUMENT NO.', 'SAMPLING RATE',]

def WCTUN1_ERR(d, errlv):
  ''' WCTUN1_ERR(2)
      Fill in remaining parameters with None when an error is encountered
      in wctun1().
      d        dictionary of all parameters
      errlv    error level to return
  '''
  for key in REQUIRED:
    if key not in d:
      d[key] = None
  return (d, errlv)

def wctun1(lst):
  global _verbose,
  warned = False
  all_items = {}

  if len(lst) == 0:
    return WCTUN1_ERR(all_items, 2)

  line = lst[0]

  if line.find('EXPOCODE') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no EXPOCODE found'
    return WCTUN1_ERR(all_items, 2)

  if line.find('WHP') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no WHPID found'
    return WCTUN1_ERR(all_items, 2)

  if line.find('EXPOCODE')+8 > line.find('WHP')-1:
    if _verbose:
      print>>sys.stderr,'wctun1: strange first header line (whpid, expo)'
    return WCTUN1_ERR(all_items, 2)
  scrtch = line[line.find('EXPOCODE')+8:line.find('WHP')+1].strip()
  all_items['EXPOCODE'] = scrtch[:scrtch.find(' ')]

  if line.find('DATE') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no DATE word found on line 1'
    return WCTUN1_ERR(all_items, 2)

  if line.find('WHP')+6 > line.find('DATE')-1:
    if _verbose:
      print>>sys.stderr,'wctun1: strange first header line (date, whpid)'
    return WCTUN1_ERR(all_items, 2)
  all_items['WHP'] = line[line.find('WHP')+6:line.find('DATE')-1].strip()

  scrtch = line[line.find('DATE')+4:line.find('DATE')+14].strip()
  cmonth, cday, cyear = -9, -9, -9

  try:
    delims = (2,4)
    if len(scrtch) == 5:
      if _verbose:
        print>>sys.stderr,'5-char date -- please fix; read i1,2i2'
      delims = (1,3)
    cmonth, cday, cyear = (int(scrtch[:delims[0]]),
                           int(scrtch[delims[0]:delims[1]]),
                           int(scrtch[delims[1]:]))
  except (TypeError, ValueError):
    if _verbose:
      print>>sys.stderr,'invalid date %s' % scrtch
    cmonth, cday, cyear = -9, -9, -9
    warned = True
  all_items['DATE'] = {'month':cmonth,'day':cday,'year':cyear} # FIXME

  line = list[1]
  if line.find('STNNBR') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no STNNBR found on line 2'
    return WCTUN1_ERR(all_items, 2)

  if line.find('CASTNO') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no CASTNO found on line 2'
    return WCTUN1_ERR(all_items, 2)

  if line.find('STNNBR')+6 > line.find('CASTNO')-1:
    if _verbose:
      print>>sys.stderr,'wctun1: strange 2nd header line (stnnbr, castno)'
    return WCTUN1_ERR(all_items, 2)
  all_items['STNNBR'] = line[line.find('STNNBR')+6:line.find('CASTNO')].strip()

  if line.find('NO.') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no NO. RECORDS= found on line 2'
    return WCTUN1_ERR(all_items, 2)

  if (line.find('NO.')-1) - (line.find('CASTNO')+6) + 1 < 1:
    if _verbose:
      print>>sys.stderr,'wctun1: strange 2nd header line (no.records, castno)'
    return WCTUN1_ERR(all_items, 2)

  try:
    all_items['CASTNO'] = int(line[line.find('NO.')-1:line.find('CASTNO')+6])
  except (TypeError, ValueError):
    if _verbose:
      print>>sys.stderr,'wctun1: invalid castno %s' % line[
          line.find('NO.')-1:line.find('CASTNO')+6]
    return WCTUN1_ERR(all_items, 2)

  if line.find('ORDS=') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no NO. RECORDS= found on line 2'
    return WCTUN1_ERR(all_items, 2)

  scrtch = line[line.find('ORDS=')+5:line.find('ORDS=')+5+10].strip()
  if scrtch.find(' ') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: NO. RECORDS= (line 2) unreadable'
    return WCTUN1_ERR(all_items, 2)

  try:
    all_items['NO. RECORDS'] = int(scrtch[:scrtch.find(' ')].strip())
  except (TypeError, ValueError):
    if _verbose:
      print>>sys.stderr,'wctun1: invalid NO. RECORDS= %s' % scrtch[
          :scrtch.find(' ')].strip()
    return WCTUN1_ERR(all_items, 2)

  line = lst[2]

  hdr3_bad = False
  if line.find('NO.') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no INSTRUMENT NO. found on line 3'
    hdr3_bad = True
  if line.find('SAMP') == -1:
    if _verbose:
      print>>sys.stderr,'wctun1: no SAMPLING RATE found on line 3'
    hdr3_bad = True
  if line.find('NO.')+3 > line.find('SAMP')-1:
    if _verbose:
      print>>sys.stderr, str('wctun1: strange 3rd header line'+
                             '(sampling rate, instrument no.)')
    hdr3_bad = True
  if hdr3_bad:
    instrument = '-'
    warned = True
  else:
    all_items['INSTRUMENT NO.'] = line[line.find('NO.')+3:
        line.find('SAMP')-1].strip()
    if len(all_items['INSTRUMENT NO.']) == 0:
      all_items['INSTRUMENT NO.'] = '-'

  if (line.find('RATE') == -1 or
      line.find('HZ')   == -1 or
     (line.find('HZ')-1) - (line.find('RATE')+4) + 1 < 2:
    err = 'unknown error with sampling rate on line 3'
    if line.find('RATE') == -1:
      err = 'no SAMPLING RATE found on line 3'
    elif line.find('HZ') == -1:
      err = 'no HZ found on line 3'
    else:
      err = "can't read sampling rate"
    if _verbose:
      print>>sys.stderr,'wctun1: %s' % err
    all_items['SAMPLING RATE'] = -9
    return WCTUN1_ERR(all_items, 1)

  try:
    unconverted = line[line.find('RATE')+4:line.find('HZ')-1]
    if (unconverted.find('.') != -1 or
        unconverted.find('e') != -1 or
        unconverted.find('E') != -1):
      all_items['SAMPLING RATE'] = float(unconverted)
    else:
      all_items['SAMPLING RATE'] = (int(unconverted[:-2])+
          float(str('.'+unconverted[-2:])))
  except (TypeError, ValueError):
    if _verbose:
      print>>sys.stderr,"wctun1: couldn't convert %s into f7.2" % unconverted
    all_items['SAMPLING RATE'] = -9
    return WCTUN1_ERR(all_items, 1)

  return (all_items, 0)
